velox/docs/functions/spark/regexp.rst

```diff
@@ -38,7 +38,12 @@ See https://github.com/google/re2/wiki/Syntax for more information.
 .. spark:function:: regexp_replace(string, pattern, overwrite) -> varchar
 
     Replaces all substrings in ``string`` that match the regular expression ``pattern`` with the string ``overwrite``. If no match is found, the original string is returned as is.
-    There is a limit to the number of unique regexes to be compiled per function call, which is 20.
+    There is a limit to the number of unique regexes to be compiled per function call, which is 20. If this limit is exceeded the function will throw an exception.
+
+    regexp_replace will throw an exception if ``string`` contains an invalid UTF-8 character, or if ``pattern`` does not conform to RE2 syntax: https://github.com/google/re2/wiki/Syntax.
+
+    regexp_replace does not support named ASCII character classes or ASCII character class union, intersection, or difference and will throw an exception if they are detected within the provided ``pattern``.
```

Comment:
Is this a limitation of RE2 or this function? Would users understand what this means? Maybe we could provide some links that explain class union, intersection,... or give some examples?

Reply:
This is a limitation of a mixture of RE2 and the way regexp_extract, rlike, and regexp_replace choose to handle that limitation.

regexp_extract and rlike call "ensureRegexIsConstantAndCompatible" which calls checkForCompatiblePattern after ensuring the pattern is constant. regexp_replace skips the constant check and directly calls checkForCompatiblePattern, which is designed to get around a limitation of RE2. 

RE2 does not allow for character class unions, intersections, or differences whereas Java does. This link can help illustrate this: https://regex101.com/r/WiVEbV/1

If you view with the "Java 8" flavor, the output should look like the following:
<img width="444" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/24e0e326-8970-43ed-ab6c-1f4014cf1bd4">
The first line is an input that is designed around Java's regex capabilities. The second line is designed around RE2. RE2, instead of viewing this as a union of character classes "a" and "bc" views it as a single character class of "a[bc" and expects a single "]" at the end. When you switch to RE2's flavor (PCRE2) you see it fully matching the second line:
<img width="434" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/f61c41fc-f176-478a-987d-d777af49cb02">


ASCII character classes (for example [[:alnum::]] ), compile and work correctly in a standalone RE2 environment, but when passed into Velox they fail the checks in checkForCompatiblePattern. 

This is because checkForCompatiblePattern was designed around basic character classes without ASCII character classes in mind. It checks for a nested opening brace and if it detects one throws an error. As shown here, ASCII character classes must contain a nested opening brace:

<img width="377" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/e8a0cf97-4dd7-40ee-a030-b237a32d2aba">

I attempted to modify the function to handle this, but as I got further and further more corner cases became clear to me and I decided that it would probably be best to just document the limitation and allow Gluten to fallback to Java instead.

In regexp_extract and rlike implementations, this limitation is only documented in the error message. I can move these warnings in the regexp_replace documentation towards to top of the file to make it clear they apply to all velox spark regexp functions.

Based on a quick search, it seems Presto also uses the Java pattern syntax with a few limitations: https://prestodb.io/docs/current/functions/regexp.html. I would assume, then, that these limitations also occur when using RE2 as a proxy for a Java regex engine. I have no experience with Presto and have nothing setup locally to check discrepancies, but taking the example function call of regexp_replace `(SELECT regexp_replace('1a 2b 14m', '(\d+)([ab]) ', '3c$2 '); -- '3ca 3cb 14m')` and replacing ([ab]) with a union, we get the following:


<img width="701" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/3bd37ec9-43e6-4506-8e04-8a786ae7f2d3">

Line 1 is the documented case working as expected (though the Velox Presto function documentation does not contain the \\$, just the $).
Line 2 is the documented case using character class union and failing.
Line 3 is the documented case adding the ] to match the way RE2 interprets the pattern and working as "expected".

Seeing the $2 in the Presto use case made me realize that that may be something that I am not accounting for in my implementation. Additionally, after rebasing I've noticed that all prod spark regexp functions implementations are now leveraging implementations in velox/functions/lib. Do you believe I should modify my implementation to do the same?

Reply:
> checkForCompatiblePattern, which is designed to get around a limitation of RE2.

I see. I feel it is very difficult and practically impossible to check the regex pattern and decide whether it will be handled by RE2 exactly the same as by Java Regex. Hence, I wonder if we should even try. I'd consider not trying and simply compiling regex and if compilation fails throw an error, otherwise proceed. I do understand that this means that query results with Gluten may be different from Spark Java, but I don't see any way to robustly protect from them, hence, suggest to clearly document that Gluten uses RE2 engine which has different syntax and semantics than Java regex.

Reply:
I agree. 

In regards to "should we even try", I spent about 2 days writing up a software suite that automates these checks on a compile and output matching basis. There are 200+ pattern types in the RE2 syntax page. Of those 200, RE2 as a direct proxy for Java regex only lacks support for ~20 patterns that Java does support.  If instead of throwing errors when a pattern fails to compile, you substitute boost, then the number drops to ~5. But, based on this [page](https://lh3lh3.users.sourceforge.net/reb.shtml) that I've tested locally, the performance degradation when using boost might not be worth it.

Reply:
Do you think we should remove the "checkForCompatiblePattern" and just clearly document the risk of differing output using Velox?

Reply:
I think so.

Reply:
> I spent about 2 days writing up a software suite that automates these checks on a compile and output matching basis

It might be helpful to summarize your findings in a blog post. CC: @zacw7 @amitkdutta

Reply:
Yeah, I'd be very interested in that. I need to go back and re-write some of it as I never planned to use it after I was done with it, so I wrote some parts directly into the Velox RegexFunctionsTest test suite (which is why I discovered the ASCII character class issue, since they do compile in the standalone environment). I deleted those when I was done in order to make a push to this PR. 

But yeah, once this PR is done I'd be glad to talk with Zac and Amit about my findings.